File: MyPreconditioners/TestAMG.py
import numpy as np
import scipy.sparse as sp
from pyamg import smoothed_aggregation_solver
import os

def load_CSC_sparse_array_from_jl(fname, n):
  colptr = np.load(fname + "-" + str(n) + "_colptr.npy") - 1
  nzval = np.load(fname + "-" + str(n) + "_nzval.npy")
  rowval = np.load(fname + "-" + str(n) + "_rowval.npy") - 1
  A = sp.csc_matrix((nzval, rowval, colptr), (n, n))
  A = sp.csr_matrix(A)
  return A

fname = "random-penta"
n = 1000

os.system("julia _TestAMG.jl")

A = load_CSC_sparse_array_from_jl(fname, n)
b = np.load("b.npy")

ml = smoothed_aggregation_solver(A)
M = ml.aspreconditioner(cycle='V')

x = np.load(fname + "-" + str(n) + "_x.npy")
res = np.linalg.norm(x - M(b))
tol = 1e-7

if res < tol:
  print("PyAMG runs properly when called from Julia.")
else:
  print("PyAMG does not run properly when called from Julia.")

os.system("rm %s-%d_colptr.npy" % (fname, n))
os.system("rm %s-%d_nzval.npy" % (fname, n))
os.system("rm %s-%d_rowval.npy" % (fname, n))
os.system("rm b.npy")
os.system("rm %s-%d_x.npy" % (fname, n))

File: MyPreconditioners/_TestAMG.jl
push!(LOAD_PATH, ".")

using Random: seed!
using NPZ: npzwrite
using LinearAlgebra: SymTridiagonal, I
using SparseArrays: sprand, sparse
using MyPreconditioners: AMGPreconditioner

seed!(1234)
const n = 1_000
const T = Float64

function get_sparse_SPD_mat(n::Int)
  A = sparse(SymTridiagonal(2 .- .05 * rand(T, n), -1 .+ .05 * rand(T, n-1)))
  A = A * A;
  return A
end

const A = get_sparse_SPD_mat(n)

function save_CSC_sparse_array_for_py(fname::String, n::Int)
  npzwrite(fname * "-" * string(n) * "_colptr.npy", A.colptr)
  npzwrite(fname * "-" * string(n) * "_nzval.npy", A.nzval)
  npzwrite(fname * "-" * string(n) * "_rowval.npy", A.rowval)
end

fname = "random-penta"

save_CSC_sparse_array_for_py(fname, n)

b = rand(T, n)
npzwrite("b.npy", b)

M = AMGPreconditioner(A)
x = M \ b
npzwrite(fname  * "-" * string(n) * "_x.npy", x)
